### PR TITLE
agent: End the agent in case of DESTROYPOD command

### PR DESCRIPTION
**Pull request by sboeuf**
We want to behave the same way hyperstart agent was behaving, that
is terminate in case a DESTROYPOD command is invoked.

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @sboeuf - I think `checkcommits` is complaining because of the colon in the `Fixes #XXX` comment. Could you remove and re-push?

**Comment by jodh-intel** [comment]
actual change...

lgtm

[Approved with PullApprove](https://pullapprove.com/clearcontainers/agent/pull-request/32/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

**Comment by jcvenegas** [comment]
Hi @sboeuf 
I am getting the following issue with 2.1
```
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.722186    2888 vm.go:114] [vm 6ac9d80e io] -> writing to hyper from #20
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.722208    2888 vm.go:121]                                     
Jun 29 23:31:55 cor-test cc-proxy[2888]: 00000000  0d                                                |.|              
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.722434    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Read from TTY
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.722489    2888 vm.go:114] [vm 6ac9d80e hyperstart]            
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.723118    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Read from tty: sequence 1 / message
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.723326    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Sequence 1 found, writing data to stdin
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.724434    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] 1 bytes written out of 1
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.724656    2888 vm.go:114] [vm 6ac9d80e io] <- writing to client #20
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.724663    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Read from stream seq 1: "\r\n"
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.724672    2888 vm.go:121]                                     
Jun 29 23:31:55 cor-test cc-proxy[2888]: 00000000  0d 0a                                             |..|             
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740037    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Stream 1 has been closed: read /dev/ptmx: input/output error
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740643    2888 vm.go:114] [vm 6ac9d80e io] <- writing to client #20
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740661    2888 vm.go:121]                                     
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740678    2888 vm.go:114] [vm 6ac9d80e io] <- writing to client #20
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740687    2888 vm.go:121]                                     
Jun 29 23:31:55 cor-test cc-proxy[2888]: 00000000  00                                                |.|              
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.740899    2888 proxy.go:81] [client #20] hyper(cmd=destroypod, data="")
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.741679    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Received "destroypod" command
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.742024    2888 vm.go:114] [vm 6ac9d80e hyperstart] ERRO[0029] Run "destroypod" command failed: 1 containers not removed, impossible to destroy the pod
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.742024    2888 proxy.go:81] [client #20] error serving client: write unix /var/run/cc-oci-runtime/proxy.sock->@: write: broken pipe
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.742056    2888 proxy.go:71] [client #20] connection closed    
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.747452    2888 proxy.go:71] [client #22] client connected     
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.747589    2888 proxy.go:81] [client #22] attach(containerId=6ac9d80efd768d73ab5170811350d114680f2545557c3a2d0fd20fddd34ef333)
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.747806    2888 proxy.go:81] [client #22] hyper(cmd=destroypod, data={})
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.748034    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Received "destroypod" command
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.748089    2888 vm.go:114] [vm 6ac9d80e hyperstart] ERRO[0029] Run "destroypod" command failed: 1 containers not removed, impossible to destroy the pod
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.748286    2888 proxy.go:71] [client #22] connection closed    
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.754268    2888 proxy.go:71] [client #23] client connected     
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.754394    2888 proxy.go:81] [client #23] attach(containerId=6ac9d80efd768d73ab5170811350d114680f2545557c3a2d0fd20fddd34ef333)
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.754662    2888 proxy.go:81] [client #23] hyper(cmd=destroypod, data={})
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.754876    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Received "destroypod" command
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.754947    2888 vm.go:114] [vm 6ac9d80e hyperstart] ERRO[0029] Run "destroypod" command failed: 1 containers not removed, impossible to destroy the pod
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.755200    2888 proxy.go:71] [client #23] connection closed    
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762139    2888 proxy.go:71] [client #24] client connected     
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762278    2888 proxy.go:81] [client #24] attach(containerId=6ac9d80efd768d73ab5170811350d114680f2545557c3a2d0fd20fddd34ef333)
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762494    2888 proxy.go:81] [client #24] hyper(cmd=destroypod, data={})
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762706    2888 vm.go:114] [vm 6ac9d80e hyperstart] INFO[0029] Received "destroypod" command
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762768    2888 vm.go:114] [vm 6ac9d80e hyperstart] ERRO[0029] Run "destroypod" command failed: 1 containers not removed, impossible to destroy the pod
Jun 29 23:31:55 cor-test cc-proxy[2888]: I0629 23:31:55.762992    2888 proxy.go:71] [client #24] connection closed    
```

**Comment by sboeuf** [comment]
@jcvenegas let me force the container removal in case of destroypod, this should fix your issue.

**Comment by jcvenegas** [comment]
@sboeuf now I get this error

```
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.455299    2888 proxy.go:81] [client #55] hyper(cmd=destroypod, data="")                                                                                                               
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.456410    2888 vm.go:114] [vm 6c594506 hyperstart] INFO[0009] Received "destroypod" command                                                                                           
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.456814    2888 proxy.go:81] [client #55] error serving client: write unix /var/run/cc-oci-runtime/proxy.sock->@: write: broken pipe                                                   
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.456858    2888 proxy.go:71] [client #55] connection closed     
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.456873    2888 vm.go:114] [vm 6c594506 hyperstart] ERRO[0009] Run "destroypod" command failed: device or resource busy                                                                
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.461744    2888 proxy.go:71] [client #57] client connected      
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.461870    2888 proxy.go:81] [client #57] attach(containerId=6c59450677b8f051fad0d20066185260e728d8230a822e7c96551bc54967bf3d)                                                         
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.462100    2888 proxy.go:81] [client #57] hyper(cmd=destroypod, data={})                                                                                                               
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.462321    2888 vm.go:114] [vm 6c594506 hyperstart] INFO[0009] Received "destroypod" command                                                                                           
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.462570    2888 vm.go:114] [vm 6c594506 hyperstart] ERRO[0009] Run "destroypod" command failed: device or resource busy                                                                
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.462777    2888 proxy.go:71] [client #57] connection closed     
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.468979    2888 proxy.go:71] [client #58] client connected      
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.469120    2888 proxy.go:81] [client #58] attach(containerId=6c59450677b8f051fad0d20066185260e728d8230a822e7c96551bc54967bf3d)                                                         
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.469358    2888 proxy.go:81] [client #58] hyper(cmd=destroypod, data={})                                                                                                               
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.469589    2888 vm.go:114] [vm 6c594506 hyperstart] INFO[0009] Received "destroypod" command                                                                                           
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.469776    2888 vm.go:114] [vm 6c594506 hyperstart] ERRO[0009] Run "destroypod" command failed: device or resource busy                                                                
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.469967    2888 proxy.go:71] [client #58] connection closed     
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.477289    2888 proxy.go:71] [client #59] client connected      
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.477420    2888 proxy.go:81] [client #59] attach(containerId=6c59450677b8f051fad0d20066185260e728d8230a822e7c96551bc54967bf3d)                                                         
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.477643    2888 proxy.go:81] [client #59] hyper(cmd=destroypod, data={})                                                                                                               
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.477871    2888 vm.go:114] [vm 6c594506 hyperstart] INFO[0009] Received "destroypod" command                                                                                           
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.478149    2888 vm.go:114] [vm 6c594506 hyperstart] ERRO[0009] Run "destroypod" command failed: device or resource busy                                                                
Jun 30 00:31:35 cor-test cc-proxy[2888]: I0630 00:31:35.478343    2888 proxy.go:71] [client #59] connection closed     
```

File: agent.go [hunk 1]
```diff
@@ -237,8 +244,12 @@ func (p *pod) streamsLoop(wg *sync.WaitGroup) {
 		seq, data, err := p.readTty()
 		if err != nil {
 			if err == io.EOF {
-				time.Sleep(time.Millisecond)
-				continue
+				select {
+				case <-p.destroyPodCh:
```
Comment (jcvenegas) [inline agent.go:248]: cold you add a info message that the ctlloop exited?

**Comment by jcvenegas** [comment]
@sboeuf  the main in the agent is doding a `wgLoops.Wait()`for two  gorutines can you add  a info message after `wgLoops.Wait()` to have an idea it exist from the wait ?

**Comment by sboeuf** [comment]
Working with Clear Containers 2.1. 3.0 has been tested and nothing is broken because of this patch. This was expected because 3.0 is not using the DESTROYPOD command.

**Pull request merged**